src/sage/rings/polynomial/polynomial_element.pyx

```diff
@@ -2617,13 +2687,18 @@ cdef class Polynomial(CommutativePolynomial):
             #       FiniteField type if the base field is a non-prime field,
             #       so this slower option is chosen to ensure the root is
             #       over explicitly a FiniteField type.
-            ring = self.base_ring().extension(f.degree(), names="a")
-
-        # Now we look for a linear root of this irreducible polynomial of degree `degree`
-        # over the user supplied ring or the extension we just computed. If the user sent
-        # a bad ring here of course there may be no root found.
-        f = f.change_ring(ring)
-        return f.any_root()
+            ring = self.base_ring().extension(degree, names="a")
+
+        # TODO: The proper thing to do here would be to call
+        #       return f.change_ring(ring).any_root()
+        #       but as we cannot work in the minimal extension (see above) working
+        #       in the extension for f.change_ring(ring).any_root() is almost always
+        #       much much much slower than using the call for roots() which uses
+        #       C library bindings for all finite fields.
+        #       Until the coercion system for finite fields works better,
+        #       this will be the most performant
+        roots = f.roots(ring, multiplicities=False)
+        return choice(roots)
 
```

Comment:
```suggestion
        return f.roots(ring, multiplicities=False)[0]

```

Reply:
Yeah I disagree with this, as Giacomo pointed out the CZ method is inherently randomised anyways (IIRC there's no competitively-fast deterministic root finding method).

Reply:
Well, the fact that a randomized algorithm is best for some cases of `.any_root()` does not imply that we should actively randomize the cases that are currently taking other code paths, too — and I agree that it's a bit arbitrary (or *random*) to do so here. As I argued in another comment, in my book it's fine for `.any_root()` to be randomized or deterministic or a little bit of both depending on the input. Eventually, forcing deterministic behaviour should probably be supported using a keyword argument, while the default would be to run the fastest algorithm that finds *any root* as the name suggests, but that's for later.